### PR TITLE
Remove "free-gpu" from *_train and create queue-freegpu.pl

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
This PR is a refactoring request of this part of *_train.py
https://github.com/espnet/espnet/blob/f6b55b38a8b5e12a7b88df887dd7d5aa5cd0f6e0/espnet/bin/asr_train.py#L313-L331,

This part works only at JHU and  such codes shouldn't be written in python level. 
I created queue-freegpu.pl, which is almost same as kaldi's `queue.pl` except that it invokes `free-gpu` command and set CUDA_VISIBLE_DEVICES as this part do.
Note that queue-freegpu.pl is a private tool for JHU. 

I also changed the behaviour of `--ngpu` of `*_train.py`:

1. If --npu is given, use as it is.
1. elif CUDA_VISIBLE_DEVICES=None, use the number of visible devices.
1. elif nvidia-smi can be invoked, use the number of all devices.
1. else ngpu==0

Recent job scheduler such as slurm, torque, can manage GPU resources and user program doesn't need to handle gpu arguments, but it is enough just to use 'visible' devices, which are allocated by Job scheduler. 

About 3. I decided to use `nvidia-smi` instead of the python method for getting the Ngpu, e.g. `torch.cuda.device_count()`, because I want not to limit the DNN backend in this part.

I also planned to remove `--ngpu` from run.sh for `*_train.py` because it is specified twice for `cmd` and there, but finally I kept it as it is.  This is need for run.pl, or ssh.pl, and I have no idea to solve it with better way. 

Actually, I don't have neither SGE nor free-gpu, so please test this tool, anyone.

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks a lot!!!
I may not find time to test it...
@27jiangziyan, could you test it at the CLSP environment?

File: egs/an4/asr1/cmd.sh [hunk 1]
```diff
@@ -80,7 +80,7 @@ elif [ "${cmd_backend}" = ssh ]; then
 elif [ "${cmd_backend}" = jhu ]; then
 
     export train_cmd="queue.pl --mem 2G"
-    export cuda_cmd="queue.pl --mem 2G --gpu 1 --config conf/gpu.conf"
+    export cuda_cmd="queue-frregpu.pl --mem 2G --gpu 1 --config conf/gpu.conf"
```
Comment (ShigekiKarita) [inline egs/an4/asr1/cmd.sh]: typo

Comment (kamo-naoyuki) [reply egs/an4/asr1/cmd.sh]: Thanks

**Comment by Xiaofei-Wang** [comment]
Hi Kamo,
I tested your pull request on the JHU cluster using --ngpu=1 but failed. Here is the returned errors:

```THCudaCheck FAIL file=/opt/conda/conda-bld/pytorch_1549635019666/work/aten/src/THC/THCGeneral.cpp line=51 error=38 : no CUDA-capable device is detected
Traceback (most recent call last):
  File "/export/c03/xwang/espnet_test1/espnet/egs/an4/asr1/../../../espnet/bin/asr_train.py", line 373, in <module>
    main(sys.argv[1:])
  File "/export/c03/xwang/espnet_test1/espnet/egs/an4/asr1/../../../espnet/bin/asr_train.py", line 361, in main
    train(args)
  File "/export/c03/xwang/espnet_test1/espnet/espnet/asr/pytorch_backend/asr.py", line 291, in train
    model = model.to(device)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 381, in to
    return self._apply(convert)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 187, in _apply
    module._apply(fn)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 187, in _apply
    module._apply(fn)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 187, in _apply
    module._apply(fn)
  [Previous line repeated 1 more time]
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/rnn.py", line 116, in _apply
    ret = super(RNNBase, self)._apply(fn)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 193, in _apply
    param.data = fn(param.data)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 379, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/export/c03/xwang/espnet_test1/espnet/tools/venv/lib/python3.7/site-packages/torch/cuda/__init__.py", line 162, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (38) : no CUDA-capable device is detected at /opt/conda/conda-bld/pytorch_1549635019666/work/aten/src/THC/THCGeneral.cpp:51```

I haven't tried the case of ngpu=None, I'll keep you posted when I finish the following tests.

**Comment by sw005320** [comment]
@Xiaofei-Wang, can you test it again?

**Comment by Xiaofei-Wang** [comment]
I've tested, it works well on JHU grid.
```# Running on c03
# Started at Wed Jul 31 00:38:04 EDT 2019
# asr_train.py --config conf/train_mtlalpha1.0.yaml --ngpu 1 --backend pytorch --outdir exp/train_nodev_pytorch_train_mtlalpha1.0/results --tensorboard-dir tensorboard/train_nodev_pytorch_train_mtlalpha1.0 --debugmode 1 --dict data/lang_1char/train_nodev_units.txt --debugdir exp/train_nodev_pytorch_train_mtlalpha1.0 --minibatches 0 --verbose 1 --resume --train-json dump/train_nodev/deltafalse/data.json --valid-json dump/train_dev/deltafalse/data.json 
free gpu: 0
2019-07-31 00:38:05,582 (asr_train:331) INFO: ngpu: 1
2019-07-31 00:38:05,583 (asr_train:334) INFO: python path = :/home/xwang/anaconda2/pkgs
2019-07-31 00:38:05,583 (asr_train:337) INFO: random seed = 1
2019-07-31 00:38:05,598 (asr_train:354) INFO: backend = pytorch
/export/c03/xwang/espnet-test/espnet/tools/venv/lib/python3.7/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
2019-07-31 00:38:15,164 (deterministic_utils:24) INFO: torch type check is disabled
2019-07-31 00:38:15,228 (asr:241) INFO: #input dims : 83
2019-07-31 00:38:15,228 (asr:242) INFO: #output dims: 30
2019-07-31 00:38:15,228 (asr:247) INFO: Pure CTC mode
2019-07-31 00:38:15,228 (e2e_asr:86) INFO: subsample: 1 2 2 1 1
2019-07-31 00:38:15,328 (encoders:236) INFO: BLSTM with every-layer projection for encoder
2019-07-31 00:38:15,572 (asr:274) INFO: writing a model config file to exp/train_nodev_pytorch_train_mtlalpha1.0/results/model.json

**Comment by sw005320** [comment]
Thanks, @Xiaofei-Wang.
@kamo-naoyuki, it would be great if you prepare `cmd.sh` for the other recipes.

**Pull request merged**